tests/unit/test_hnsw_tiered.cpp

```diff
@@ -1487,3 +1432,280 @@ TYPED_TEST(HNSWTieredIndexTest, parallelInsertAdHoc) {
+
+        thread_pool_wait(jobQ, run_thread);
+
+        EXPECT_EQ(tiered_index->getHNSWIndex()->checkIntegrity().connections_to_repair, 0);
+        EXPECT_EQ(tiered_index->getHNSWIndex()->safeGetEntryPointCopy(), INVALID_ID);
+        // Verify that we have no pending jobs.
+        EXPECT_EQ(tiered_index->labelToInsertJobs.size(), 0);
+        EXPECT_EQ(tiered_index->idToRepairJobs.size(), 0);
+        EXPECT_EQ(tiered_index->idToSwapJob.size(), tiered_index->swapJobs.size());
+        for (auto job : tiered_index->swapJobs) {
+            EXPECT_EQ(job->pending_repair_jobs_counter.load(), 0);
+        }
+    }
+}
+
+TYPED_TEST(HNSWTieredIndexTest, alternateInsertDeleteAsync) {
+    std::shared_ptr<VecSimAllocator> allocator = VecSimAllocator::newVecsimAllocator();
+
+    // Create TieredHNSW index instance with a mock queue.
+    size_t dim = 16;
+    size_t n = 1000;
+    for (auto isMulti : {false, true}) {
+        HNSWParams params = {.type = TypeParam::get_index_type(),
+                             .dim = dim,
+                             .metric = VecSimMetric_L2,
+                             .multi = isMulti,
+                             .blockSize = 100};
+        auto jobQ = JobQueue();
+        auto index_ctx = IndexExtCtx();
+        size_t memory_ctx = 0;
+        TieredIndexParams tiered_params = {.jobQueue = &jobQ,
+                                           .jobQueueCtx = &index_ctx,
+                                           .submitCb = submit_callback,
+                                           .memoryCtx = &memory_ctx,
+                                           .UpdateMemCb = update_mem_callback};
+        TieredHNSWParams tiered_hnsw_params = {.hnswParams = params, .tieredParams = tiered_params};
+        auto *tiered_index = reinterpret_cast<TieredHNSWIndex<TEST_DATA_T, TEST_DIST_T> *>(
+            HNSWFactory::NewTieredIndex(&tiered_hnsw_params, allocator));
+        // Set the created tiered index in the index external context.
+        index_ctx.index_strong_ref.reset(tiered_index);
+
+        size_t per_label = isMulti ? 10 : 1;
+        size_t n_labels = n / per_label;
+
+        // Launch the BG threads loop that takes jobs from the queue and executes them.
+        bool run_thread = true;
+        for (size_t i = 0; i < THREAD_POOL_SIZE; i++) {
+            thread_pool.emplace_back(thread_main_loop, std::ref(jobQ), std::ref(run_thread));
+        }
+
+        // Create and insert 10 vectors, then delete them right after.
+        size_t batch_size = 10;
+        std::srand(10); // create pseudo random generator with any arbitrary seed.
+        for (size_t i = 0; i < n / batch_size; i++) {
+            for (size_t l = 0; l < batch_size; l++) {
+                TEST_DATA_T vector[dim];
+                for (size_t j = 0; j < dim; j++) {
+                    vector[j] = std::rand() / (TEST_DATA_T)RAND_MAX;
+                }
+                tiered_index->addVector(vector, (i * batch_size + l) % n_labels);
```

Comment:
this formula guarantees (with the given parameters) that no 2 vectors with the same label will be inserted and deleted in the same batch. consider having another formula